### PR TITLE
Fix incorrect date validation from dayjs

### PR DESCRIPTION
**Pull request by RCout1nho**
## What is the problem?
As shown at https://github.com/mui/material-ui/issues/30886 and https://github.com/mui/material-ui/issues/29838, there is a problem validating dates using dayjs.
I noticed that the [parse](https://github.com/dmtrKovalenko/date-io/blob/dc9a2ba540f6032fab5d40d57ba646d8d25e33ba/packages/dayjs/src/dayjs-utils.ts#L108) function uses dayjs parse without strict format, causing dayjs to convert any date, even if incorrect, e.g. "99/99/2000" to a correct date, and due to this, all other functions used later ([date](https://github.com/dmtrKovalenko/date-io/blob/dc9a2ba540f6032fab5d40d57ba646d8d25e33ba/packages/dayjs/src/dayjs-utils.ts#L116), [isValid](https://github.com/dmtrKovalenko/date-io/blob/dc9a2ba540f6032fab5d40d57ba646d8d25e33ba/packages/dayjs/src/dayjs-utils.ts#L128), etc) do not have ability to validade dates, as they will always receive a supposedly valid date.

This issue is the same as #599

## What did I do?
* I changed the type of the `local` [here](https://github.com/dmtrKovalenko/date-io/blob/dc9a2ba540f6032fab5d40d57ba646d8d25e33ba/packages/dayjs/src/dayjs-utils.ts#L12) and [here](https://github.com/dmtrKovalenko/date-io/blob/dc9a2ba540f6032fab5d40d57ba646d8d25e33ba/packages/dayjs/src/dayjs-utils.ts#L62) from string to `ILocale`, because it's the type passed by dayjs
* Due to the previous change, I changed the places where `this.locale` was called as string to `this.locale?.name`, because it's the equivalent
* I changed [parse](https://github.com/dmtrKovalenko/date-io/blob/dc9a2ba540f6032fab5d40d57ba646d8d25e33ba/packages/dayjs/src/dayjs-utils.ts#L108) function return from `this.dayjs(value, format, this.locale)` to `this.dayjs(value, format, this.locale?.name, true)`, enabling strict mode.

## Note
I'm working with material-ui DatePickers, and with my changes, pickers work perfectly

### CODE REVIEW

**Comment by dmtrKovalenko** [comment]
The biggest question from me – will it require breaking change? If not - I will be happy to merge it and release right now

File: packages/dayjs/src/dayjs-utils.ts [hunk 1]
```diff
@@ -57,17 +57,21 @@ const defaultFormats: DateIOFormats = {
 
 export default class DayjsUtils<TDate extends Dayjs = Dayjs> implements IUtils<TDate> {
   public rawDayJsInstance: typeof defaultDayjs;
+
   public lib = "dayjs";
+
   public dayjs: Constructor<TDate>;
+
   public locale?: string;
+
```
Comment (dmtrKovalenko) [inline packages/dayjs/src/dayjs-utils.ts:66]: Please remove unnecessary changes

File: packages/dayjs/src/dayjs-utils.ts [hunk 2]
```diff
@@ -81,11 +85,11 @@ export default class DayjsUtils<TDate extends Dayjs = Dayjs> implements IUtils<T
...
-    var localFormattingTokens = /(\[[^\[]*\])|(\\)?(LTS|LT|LL?L?L?)|./g;
+    const localFormattingTokens = /(\[[^\[]*\])|(\\)?(LTS|LT|LL?L?L?)|./g;
     return format
       .match(localFormattingTokens)
       .map((token) => {
-        var firstCharacter = token[0];
+        const firstCharacter = token[0];
         if (firstCharacter === "L") {
           /* istanbul ignore next */
           return this.rawDayJsInstance.Ls[this.locale || "en"]?.formats[token] ?? token;
```
Comment (dmtrKovalenko) [inline packages/dayjs/src/dayjs-utils.ts:95]: Please remove unnecessary changes

**Review by dmtrKovalenko — CHANGES_REQUESTED** [review]

File: packages/dayjs/src/dayjs-utils.ts (hunk 1 above)
Comment (RCout1nho) [reply packages/dayjs/src/dayjs-utils.ts]: sorry about that, vscode changed this and I didn't even notice

**Comment by RCout1nho** [comment]
> The biggest question from me – will it require breaking change? If not - I will be happy to merge it and release right now

No, it does not require a breaking change. In fact, after this change, date-io/dayjs will have the same behavior as date-io/datefns, date-io/moment and others, as it will correctly deliver the validated dates and indicating if there are errors.
Do you agree?

**Comment by RCout1nho** [comment]
I even noticed that the LocatizationProvider component of material-ui was causing some of the confusion, as it allows the locale to be passed both as a string and as an object. However, date-io always accepted only as a string, and this caused an error when validating the locale, and made it always use the default locale ("en"). I will report this to the material-ui people.

**Comment by chwallen** [comment]
> I even noticed that the `LocatizationProvider` component of `material-ui` was causing some of the confusion, as it allows the locale to be passed both as a string and as an object. However, date-io always accepted only as a string, and this caused an error when validating the locale, and made it always use the default locale ("en"). I will report this to the material-ui people.

Yes, it was discussed in #588.

**Comment by chwallen** [comment]
With your reverts, this PR is exactly the same as #599.

**Comment by kjellski** [comment]
Can you guys merge either #599 OR #608 ? Changes are equal... but the fix would be great... :smiley:  @dmtrKovalenko

**Comment by kjellski** [comment]
:heart:  @dmtrKovalenko awesome! :+1:  thanks a lot!
